### ISSUE TITLE
Get list of all available analyzers. Request for a new _analyze API?

### ISSUE THREAD

**Issue by lukas-vlcek** [opensearch-project/OpenSearch#5481]
**Is your feature request related to a problem? Please describe.**

I am missing an option to get list of all available analyzers.
There is already analyze API documentation and it mentions "built-in" analyzers. But for normal user there is no way how to learn what are all the options. And even for people who are familiar with the code the list is a subject to updates/changes.
One option would be to document the list in the documentation page. But I think I would prefer if the cluster itself can give a list of its analyzers (and also tokenizers, charfilters, normalizers).

- There are built-in analyzers at the cluster level
- There are also ad-hoc/custom analyzers at individual index level (but visibility of these should be subject of RBA rules?)

**Describe the solution you'd like**

As far as I understand list of all built-in analyzers is materialized once `AnalysisModule.setupAnalyzers(plugins)` is called.
I think it would be useful to extend the `_analyze/` API and give it option to return list of all built-in analyzers (and its componenets: tokenizers, ... etc).

As for the list of analyzers defined at the index level I am not sure at this point. Maybe later... 

**Describe alternatives you've considered**

Alternative is to go to the documentation (which does not have this list) or go to the code (which is not an option for many people).

**Additional context**
n/a

**Comment by dtaivpp** [comment]
@lukas-vlcek I couldn't agree more. What may even be interesting is if we could start to expose some of these through a generated "cluster documentation" page in dashboards. That could show a bit of the clusters meta information.

**Comment by lukas-vlcek** [comment]
BTW, I am looking at this and I am trying to implement a quick prototype. Feel free to assign me.

**Comment by dtaivpp** [comment]
@lukas-vlcek This looks slick! I will try it out and see about getting it socialized a bit so we can have some feedback.

**Comment by lukas-vlcek** [comment]
FYI, if you will be testing the 1.0.0-rc.1 plugin then please be aware of some [known issues](https://github.com/lukas-vlcek/OpenSearch-list-built-in-analyzers/issues) and also some [fixes that are not included](https://github.com/lukas-vlcek/OpenSearch-list-built-in-analyzers/issues?q=is%3Aissue+is%3Aclosed) in that release. Of course you can always build the plugin from the source...

**Comment by dtaivpp** [comment]
I am just building myself. 2 questions:

1. @lukas-vlcek  can you present this on 1/17 https://forum.opensearch.org/t/opensearch-community-meeting-2023-0117/11891

2. Checking my understanding here. When I create a new index template with an analyzer should that show up in this list or is it just for core analyzers?

Example: 
```json
PUT _template/twitter
{
            "index_patterns": [
                "twitter*"
            ],
            "template": {
                "settings": {
                    "analysis": {
                        "analyzer": {
                            "text_analyzer": {
                                "tokenizer": "standard",
                                "filter": [ "stop" ]
                            }
                        }
                    }
                },
                "mappings": {}
        }
    }
```

Here I was thinking `text_analyzer` would show in the list but it wasn't from what I could tell. I queried as both privileged and unprivileged users.

**Comment by lukas-vlcek** [comment]
@dtaivpp 

1. Yes, I am glad to present this work. Feel free to include me, I am already signed up for the meeting.
2. The plugin pulls the list of all the "build-in" analysis components. That means the **text analysis building blocks** available to all users (and these are defined at the node level). It is by definition a static list, it does not changes during the life of the cluster (besides cluster rolling upgrade which can bring in a new version of OpenSearch or add another AnalysisPlugin). This suggested new API is primarily meant to provide complementary information for the documentation (or as you pointed out earlier, "generated documentation").  On the other hand, the custom/ad-hoc analyzer components defined at the index level is a different thing. These can change frequently and they live at the index level (so maybe `/_index/analyzers` would be more appropriate end point for such information). And the biggest difference is that they are not available to all users, if the user does not see the index (does not have the privs to see it) then he/she should not see such analyzer components as well (some sensitive info could leak this way). Other users can not "re-use" there analyzers as well, they always have to recreate them on their own indices, ... etc.

**Comment by andrross** [comment]
My two cents here is that implementing this as a core component is the right way to go architecturally. Unfortunately I missed the January 17 community meeting but is there any additional feedback to incorporate here as to the structure of the API itself?

**Comment by lukas-vlcek** [comment]
@andrross I think the best place to provide feedback about this functionality is here, in this ticket. I am going to release a new RC version because some issues about output format has been fixed in the meantime, and OpenSearch 2.5 has been released as well.

Yes, I agree the best way would be to integrate it directly into the core. But as a proof of concept it was easier for me to implement it as a plugin because I did not have to care much about frequent changes happening in `main` branch.

As for the output format I remember one feedback was that this information could be part of the `_cat` API as oppose to introducing a new REST API. I liked this idea initially but now I do not think that would be a good fit, mostly because I can not think of good response format for the `_cat` API.

**Comment by andrross** [comment]
> I think the best place to provide feedback about this functionality is here, in this ticket.

Agreed! I was just asking to capture any feedback from the meeting into this ticket :)

> ...could be part of the `_cat` API

Yeah it does seem to be hard to model the structured/nested data in this API in the CAT format. On that front though, the large JSON response payload isn't the most human-readable format, so some sort of admin UI would seem to be a good fit here.

**Comment by lukas-vlcek** [comment]
@macohen If there are any questions, feel free to ping me, I am happy to help.

**Comment by macohen** [comment]
@lukas-vlcek are you planning to keep going on this? I would encourage that! Mostly I brought it into the Search Applications Vertical project because there's some alignment there in other ways.

**Comment by macohen** [comment]
@andrross do you think the admin UI is required to launch this? @lukas-vlcek, are you able to turn this into a core component? I agree with Andrew because some analyzers are in core already.

**Comment by lukas-vlcek** [comment]
@macohen Making it a core component is perfectly possible and will make implementation a little bit more transparent/clean.

**Comment by msfroh** [comment]
I just wanted to call out a related behavior that I just learned about for pipelines and processor plugins. 

The `NodeInfo` class has a [field of type `IngestInfo`](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/action/admin/cluster/node/info/NodeInfo.java#L116) that keeps track of ingest processors available on every node. When a new ingest pipeline is created, the node that receives the request fetches all of the `NodeInfo`s, and confirms that every processor used in the pipeline is available on every node (to avoid a situation where some nodes fail to run the pipeline). I just [added similar logic for search pipelines](https://github.com/opensearch-project/OpenSearch/pull/6587), since I copied the idea from ingest pipelines.

It feels like we would have a similar situation with analyzers, where you could specify an analyzer chain for a given field in your mapping, but it would only work reliably if every component of the chain is available on every node. @lukas-vlcek, do you know off-hand how mappings accomplish that? 

I'm wondering if there might be some opportunity to make the implementations more consistent between analyzers and processor pipelines (either putting everything into `NodeInfo`, so it all gets returned via the `/_nodes` API, or we could move the pipeline processor info into a sub-API like you've done here for analyzers, making `NodeInfo` a little smaller).

**Comment by lukas-vlcek** [comment]
@msfroh Thanks for looking at this. I am going to look at that.

**Comment by macohen** [comment]
Checking in @lukas-vlcek... need anything on this one?

**Comment by lukas-vlcek** [comment]
Pushed first draft PR.
Except for some more tests it should be ready for review.
The documentation PR is missing at this point (I will open it quickly).

**Comment by macohen** [comment]
@krishna-ggk can you please take a look at this issue and the associated PR?

@lukas-vlcek, do you think this is still on track for 2.12? Not a problem to move to 2.13, but just wanted to make sure we get everything aligned for 2.12 if this is still good to go.

Thanks!

**Comment by macohen** [comment]
@lukas-vlcek Jan 9th is the current entry window for 2.12. If we work back from there, I'd say having the PR merged by Jan 7th would be a good target to know if this will make it. Should we move to 2.13?

**Comment by lukas-vlcek** [comment]
@macohen Sorry for not replying sooner, I will do my best to make it by Jan 7th. It is on my priority list now.

**Comment by kiranprakash154** [comment]
Hi, are we on track for this to be released in 2.12 ?

**Comment by lukas-vlcek** [comment]
Hi @kiranprakash154, depends on when is the code freeze for 2.12 and if we get more reviews on this PR. I am currently finishing documentation PR, I will push it on Monday.

**Comment by macohen** [comment]
Code freeze for 2.12 is Feb 6th. @lukas-vlcek do you need any assistance to get this in?

**Comment by lukas-vlcek** [comment]
@macohen
- The documentation part has been already reviewed and approved (https://github.com/opensearch-project/documentation-website/pull/6252).
- The code part is (IMO) suffering from flaky tests and still needs more reviews (https://github.com/opensearch-project/OpenSearch/pull/10296).

**Comment by hdhalter** [comment]
@macohen - Can we please bump this up to release train 2.13?

**Comment by hdhalter** [comment]
> @macohen - Can we please bump this up to release train 2.13?

Since this is still on the 2.13 roadmap, I'll move it to the 2.13 release train in the project.

**Comment by hdhalter** [comment]
Hi @lukas-vlcek, We closed the doc issue for List Analyzers Through _cat" (https://github.com/opensearch-project/documentation-website/issues/5426), but I don't think there was a doc issue for this one, specifically. Are we releasing this in 2.13 and will it need documentation? Thanks!

**Comment by dblock** [comment]
https://github.com/opensearch-project/OpenSearch/pull/10296 is next to be merged, then we can update documentation for 2.13 accordingly.

**Comment by getsaurabh02** [comment]
@lukas-vlcek should I update the tag to 2.15?
